[dynamo, nested graph breaks] support very simple nested graph breaks

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160611
* #160601
* #160470
* #159786
* #160138
* #159817
* #159678
* __->__ #159329

e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


@torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

williamwen42 committed:
[dynamo, nested graph breaks] support very simple nested graph breaks

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/resume_execution.py:
```diff
@@ -462,15 +466,74 @@ def update(
+                        # in the nested resume
+                        create_instruction("LOAD_CONST", argval=False),
+                        create_instruction(
+                            "STORE_FAST", argval=IS_TRACING_RESUME_PROLOGUE_VARNAME
```
Can you just write a comment about this one? I am not sure why we need this flag.

williamwen42 replied:
This is used to prevent dynamo from graph breaking in resume function prologues (defined [here](https://github.com/pytorch/pytorch/pull/159329/files#diff-957b95045ab713eecb0b16216229756cb2e314018c21db825ab01a269eeea3a1R54), set [here](https://github.com/pytorch/pytorch/pull/159329/files#diff-957b95045ab713eecb0b16216229756cb2e314018c21db825ab01a269eeea3a1R403), unset [here](https://github.com/pytorch/pytorch/pull/159329/files#diff-957b95045ab713eecb0b16216229756cb2e314018c21db825ab01a269eeea3a1R521))

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

atalman commented:
@pytorchmergebot revert -c ghfirst -m "failing internally"

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] support very simple nested graph breaks"


e.g. this graph breaks once now:
```python
import torch

torch._dynamo.config.nested_graph_breaks = True


def inner(x):
    x = x + 1
    torch._dynamo.graph_break()
    return x + 2


torch.compile(backend="eager")
def outer(x):
    return inner(x)


print(outer(torch.ones(3)))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
